### DISCUSSION TITLE
How to properly use `@filter(op: "regex")`?

### DISCUSSION THREAD

**Discussion by elenakrittik** [obi1kenobi/trustfall#587] [Q&A] [answered]
From [docs](https://docs.rs/trustfall_core/latest/trustfall_core/ir/enum.Operation.html) and [other discussions](https://github.com/obi1kenobi/trustfall/discussions/225), it looks like `@filter(op: "regex", values: [...])` expects a single value (the regex pattern), but [when i try to do the same](https://github.com/elenakrittik/gdtk/blob/feat/trustfall/crates/gdtk-lint/lints/style/non-pascalcase-class-name.ron), `trustfall::execute_query()` results in:

```
thread 'main' panicked at crates\gdtk-lint\src\lib.rs:15:14:
called `Result::unwrap()` on an `Err` value: Unexpected error: Filter argument count mismatch: expected 1 but found 0
```

..which suggests that there wasn't any value passed to it. However, through trial and error i noticed that removing the filter but leaving `args` intact results in this:

```
thread 'main' panicked at crates\gdtk-lint\src\lib.rs:15:14:
called `Result::unwrap()` on an `Err` value: One or more of the provided arguments are not used in this query: ["pascalcase_regex"]
```

..which, contrary to the previous message, implies that the removed filter *did* actually "use" the regex value.

Hence the question: what is the proper way to use `@filter(op: "regex")`? I looked everywhere i can but the discussion linked at the top is the only mention of regex filters, and the OP didn't provide a repository. I'll only be able to try and create a MRE tomorrow so this is just a question until then, but if you have the desire to look into this right away, i tried to make things as simple as possible:

- To reproduce, clone [the repo](https://github.com/elenakrittik/gdtk/), checkout `feat/trustfall` and run `cargo run -p gdtk lint quick.gd`.
- All of the trustfall-related code is in `crates/gdtk-lint/`, including the schema, the adapter, and the queries themselves.
  - In order to "embed" query files into the executable i used a custom `build.rs` that produces `crates/gdtk-lint/src/builtin.rs`.
  - The entrypoint that runs all queries is `run_builtin_lints()` located in `lib.rs`.
  - The adapter is located in `crates\gdtk-lint\src\adapter\`.

Thanks in advance!

**Comment by obi1kenobi** [comment, marked as answer]
I think there's a minor typo in the query that means the argument was never _actually_ passed: https://github.com/elenakrittik/gdtk/pull/49

The query doesn't actually validate against the schema with this typo, since the `@filter` directive doesn't take a `values` key per its schema definition. But the query parser on Trustfall's side was a little too permissive here and didn't raise that as an error. I'll open an issue for that — would you be open to putting together a PR to fix that by any chance?

**Reply by obi1kenobi** [reply]
It was actually an easy enough fix and touched on a couple other issues I've been meaning to fix, so I've merged a couple of PRs to improve things around this going forward: #588 and #589.

**Reply by elenakrittik** [reply]
Oh, so i overlooked a typo? I must really sleep more 😅 Thanks a lot for a quick reply & fix!
